File: Components/DevExpressVCL/Demos/ExpressScheduler/Delphi/CustomDrawDemo/CustomDrawDemoMain.pas
unit CustomDrawDemoMain;

interface

{$I cxVer.inc}

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Types, 
  Dialogs, DemoBasicMain, cxClasses, cxStyles, cxGraphics, cxEdit, cxLookAndFeelPainters, cxScheduler,
  cxSchedulerCustomControls, cxSchedulerCustomResourceView,
  cxSchedulerDayView, cxSchedulerDateNavigator, cxLookAndFeels, Menus,
  StdCtrls, cxControls, cxSchedulerStorage, ExtCtrls, ComCtrls,
  cxSchedulerTimeGridView, cxSchedulerUtils, cxSchedulerWeekView,
  cxSchedulerYearView, cxGeometry, cxSchedulerAgendaView, cxSchedulerHolidays,
  cxSchedulerGanttView;

type
  TCustomDrawDemoMainForm = class(TDemoBasicMainForm)
    Storage: TcxSchedulerStorage;
    CustomDraw1: TMenuItem;
    miEvents: TMenuItem;
    miTimeRuler: TMenuItem;
    miHeaders: TMenuItem;
    miContent: TMenuItem;
    DateNavigator1: TMenuItem;
    miMonthHeaders: TMenuItem;
    miDayCaptions: TMenuItem;
    miDays: TMenuItem;
    ViewDay1: TMenuItem;
    miContainer: TMenuItem;
    miResources: TMenuItem;
    miGroupSeparator: TMenuItem;
    cxStyleRepository1: TcxStyleRepository;
    csBoldItalic: TcxStyle;
    csItalic: TcxStyle;
    csRed: TcxStyle;
    miDNContent: TMenuItem;
    procedure FormCreate(Sender: TObject);
    procedure SchedulerDateNavigatorCustomDrawHeader(Sender: TObject;
      ACanvas: TcxCanvas;
      AViewInfo: TcxSchedulerDateNavigatorMonthHeaderViewInfo;
      var ADone: Boolean);
    procedure SchedulerDateNavigatorCustomDrawDayCaption(Sender: TObject;
      ACanvas: TcxCanvas;
      AViewInfo: TcxSchedulerDateNavigatorDayCaptionViewInfo;
      var ADone: Boolean);
    procedure SchedulerDateNavigatorCustomDrawDayNumber(Sender: TObject;
      ACanvas: TcxCanvas;
      AViewInfo: TcxSchedulerDateNavigatorDayNumberViewInfo;
      var ADone: Boolean);
    procedure SchedulerDateNavigatorCustomDrawContent(Sender: TObject;
      ACanvas: TcxCanvas;
      AViewInfo: TcxSchedulerDateNavigatorMonthContentViewInfo;
      var ADone: Boolean);
    procedure SchedulerCustomDrawDayHeader(Sender: TObject;
      ACanvas: TcxCanvas; AViewInfo: TcxSchedulerDayHeaderCellViewInfo;
      var ADone: Boolean);
    procedure SchedulerViewDayCustomDrawRuler(Sender: TObject;
      ACanvas: TcxCanvas; AViewInfo: TcxSchedulerTimeRulerCellViewInfo;
      var ADone: Boolean);
    procedure SchedulerViewDayCustomDrawContainer(Sender: TObject;
      ACanvas: TcxCanvas; AViewInfo: TcxSchedulerContainerCellViewInfo;
      var ADone: Boolean);
    procedure SchedulerCustomDrawEvent(Sender: TObject;
      ACanvas: TcxCanvas; AViewInfo: TcxSchedulerEventCellViewInfo;
      var ADone: Boolean);
    procedure SchedulerCustomDrawContent(Sender: TObject;
      ACanvas: TcxCanvas; AViewInfo: TcxSchedulerContentCellViewInfo;
      var ADone: Boolean);
    procedure SchedulerCustomDrawResourceHeader(Sender: TObject;
      ACanvas: TcxCanvas; AViewInfo: TcxSchedulerHeaderCellViewInfo;
      var ADone: Boolean);
    procedure SchedulerCustomDrawGroupSeparator(Sender: TObject;
      ACanvas: TcxCanvas;
      AViewInfo: TcxSchedulerGroupSeparatorCellViewInfo;
      var ADone: Boolean);
    procedure FormDestroy(Sender: TObject);
    procedure UpdateCustomDraw(Sender: TObject);
  private
    FDaysOffIndexes: array[1..2] of Byte;
  end;

var
  CustomDrawDemoMainForm: TCustomDrawDemoMainForm;

implementation

uses
  cxDateUtils, DateUtils,dxOffice11, cxFormats;

{$R *.dfm}

var
  W1, H1, C1, C2: Integer;
  ABitmap: TBitmap;

procedure FillGradientRect(ACanvas: TcxCanvas;
  const ARect: TRect; AColor1, AColor2: TColor; AHorizontal: Boolean);
begin
  with ARect do
  begin
    W1 := Right - Left;
    H1 := Bottom - Top;
    if (W1 < 1) or (H1 < 1) then Exit;
    if (ABitmap.Height <> H1) or (ABitmap.Width <> W1) or (AColor1 <> C1) or (AColor2 <> C2) then
    begin
      C1 := AColor1;
      C2 := AColor2;
      ABitmap.Free;
      ABitmap := TBitmap.Create;
      ABitmap.PixelFormat := pf32Bit;
      ABitmap.Height := H1;
      ABitmap.Width := W1;
      Office11FillTubeGradientRect(ABitmap.Canvas.Handle, Rect(0, 0, W1, H1), AColor1, AColor2, AHorizontal);
    end;
    BitBlt(ACanvas.Handle, Left, Top, W1, H1, ABitmap.Canvas.Handle, 0, 0, srcCopy);
  end;
end;

procedure TCustomDrawDemoMainForm.FormCreate(Sender: TObject);
var
  I, J: Integer;
begin
  inherited FormCreate(Sender);
  ABitmap := TBitmap.Create;
  ABitmap.PixelFormat := pf32bit;
  Scheduler.OptionsView.RotateResourceCaptions := False;
  Scheduler.ViewDay.EventShadows := False;
  I := Ord(cxFormatController.StartOfWeek) + Ord(dSunday);
  J := Ord(dSunday) - I;
  if J < 0 then Inc(J, 7);
  FDaysOffIndexes[1] := J;
  J := Ord(dSaturday) - I;
  if J < 0 then Inc(J, 7);
  FDaysOffIndexes[2] := J;
  GenerateRandomEvents(100, True);
end;

procedure TCustomDrawDemoMainForm.SchedulerDateNavigatorCustomDrawHeader(
  Sender: TObject; ACanvas: TcxCanvas;
  AViewInfo: TcxSchedulerDateNavigatorMonthHeaderViewInfo;
  var ADone: Boolean);
begin
  if not miMonthHeaders.Checked then Exit;
  FillGradientRect(ACanvas, AViewInfo.Bounds, $FFE0E0, $FF8080, False);
  ACanvas.FrameRect(AViewInfo.Bounds, clBlue);
  AViewInfo.Transparent := True;
end;

procedure TCustomDrawDemoMainForm.SchedulerDateNavigatorCustomDrawDayCaption(
  Sender: TObject; ACanvas: TcxCanvas;
  AViewInfo: TcxSchedulerDateNavigatorDayCaptionViewInfo;
  var ADone: Boolean);
begin
  if not miDayCaptions.Checked then Exit;
  if AViewInfo.Index in [FDaysOffIndexes[1], FDaysOffIndexes[2]] then
    ACanvas.Font := csRed.Font;
end;

procedure TCustomDrawDemoMainForm.SchedulerDateNavigatorCustomDrawDayNumber(
  Sender: TObject; ACanvas: TcxCanvas;
  AViewInfo: TcxSchedulerDateNavigatorDayNumberViewInfo;
  var ADone: Boolean);
begin
  if not miDays.Checked then Exit;
  if AViewInfo.Selected then
    ACanvas.Brush.Color := clAppWorkSpace
  else
    if DayOfWeek(AViewInfo.Date) in [1, 7] then
      ACanvas.Font := csRed.Font;
end;

procedure TCustomDrawDemoMainForm.SchedulerDateNavigatorCustomDrawContent(
  Sender: TObject; ACanvas: TcxCanvas;
  AViewInfo: TcxSchedulerDateNavigatorMonthContentViewInfo;
  var ADone: Boolean);
var
  AColor: TColor;
  R: TRect;
begin
  if not miDNContent.Checked then Exit;
  R := AViewInfo.Bounds;
  case AViewInfo.Month of
    3..5: AColor := $D0FFD0;
    6..8: AColor := $D0D0FF;
    9..11: AColor := $D0FFFF;
  else
    AColor := $FFE7E7;
  end;
  with ACanvas do
  begin
    Brush.Color := AColor;
    FillRect(R);
    Font.Height := R.Bottom - R.Top;
    Font.Color := dxGetMiddleRGB(AColor, 0, 85);
    DrawText(IntToStr(AViewInfo.Month), R, cxAlignCenter);
  end;
  ACanvas.Font := AViewInfo.ViewParams.Font;
  AViewInfo.Transparent := True;
  ADone := True;
end;

procedure TCustomDrawDemoMainForm.SchedulerCustomDrawDayHeader(
  Sender: TObject; ACanvas: TcxCanvas;
  AViewInfo: TcxSchedulerDayHeaderCellViewInfo; var ADone: Boolean);
var
  ADisplayText: string;
  ABounds: TRect;
  P: Integer;
begin
  if not miHeaders.Checked then Exit;
  AViewInfo.Transparent := True;
  FillGradientRect(ACanvas, AViewInfo.Bounds, $A0A0A0, $707070, False);
  ACanvas.FrameRect(AViewInfo.Bounds, clGray);
  ACanvas.Brush.Style := bsClear;
  ACanvas.Font := csItalic.Font;
  if (AViewInfo is TcxSchedulerAgendaViewDayHeaderCellViewInfo) and
    (Scheduler.ViewAgenda.DayHeaderOrientation = dhoVertical) then
  begin
    ADisplayText := AViewInfo.DisplayText;
    P := Pos(',', ADisplayText);
    ADisplayText := Copy(ADisplayText, 1, P - 1) + #13#10 + Copy(ADisplayText, P + 2, Length(ADisplayText) - P - 1);
    ABounds := cxRectInflate(AViewInfo.Bounds, -2 * cxTextOffset, -2 * cxTextOffset);
    ACanvas.DrawText(ADisplayText, ABounds, cxAlignHCenter or cxAlignTop or cxWordBreak);
  end
  else
    ACanvas.DrawText(AViewInfo.DisplayText, AViewInfo.Bounds, cxAlignCenter);
  ACanvas.Brush.Style := bsSolid;
  ADone := True;
end;

procedure TCustomDrawDemoMainForm.SchedulerViewDayCustomDrawRuler(
  Sender: TObject; ACanvas: TcxCanvas;
  AViewInfo: TcxSchedulerTimeRulerCellViewInfo; var ADone: Boolean);
begin
  if not miTimeRuler.Checked then Exit;
  AViewInfo.Transparent := True;
  FillGradientRect(ACanvas, AViewInfo.Bounds[True], $909090, $E0E0E0, True);
end;

procedure TCustomDrawDemoMainForm.SchedulerViewDayCustomDrawContainer(
  Sender: TObject; ACanvas: TcxCanvas;
  AViewInfo: TcxSchedulerContainerCellViewInfo; var ADone: Boolean);
begin
  if not miContainer.Checked or AViewInfo.Selected then Exit;
  AViewInfo.Transparent := True;
  FillGradientRect(ACanvas, AViewInfo.Bounds, $C0E0C0, $70A070, False);
end;

procedure TCustomDrawDemoMainForm.SchedulerCustomDrawEvent(
  Sender: TObject; ACanvas: TcxCanvas;
  AViewInfo: TcxSchedulerEventCellViewInfo; var ADone: Boolean);
begin
  if not miEvents.Checked then Exit;
  AViewInfo.Transparent := True;
  FillGradientRect(ACanvas, AViewInfo.Bounds, clWhite,
    AViewInfo.Color, False);
end;

procedure TCustomDrawDemoMainForm.SchedulerCustomDrawContent(
  Sender: TObject; ACanvas: TcxCanvas;
  AViewInfo: TcxSchedulerContentCellViewInfo; var ADone: Boolean);
begin
  if not miContent.Checked then Exit;
  AViewInfo.Transparent := True;
  if AViewInfo is TcxSchedulerAgendaViewFreeDayCellViewInfo then
    FillGradientRect(ACanvas, AViewInfo.Bounds, clWhite, AViewInfo.Color, False)
  else
    FillGradientRect(ACanvas, AViewInfo.Bounds, AViewInfo.Color, dxGetMiddleRGB(AViewInfo.Color, 0, 75), True);
end;

procedure TCustomDrawDemoMainForm.SchedulerCustomDrawResourceHeader(
  Sender: TObject; ACanvas: TcxCanvas;
  AViewInfo: TcxSchedulerHeaderCellViewInfo; var ADone: Boolean);
begin
  if not miResources.Checked then Exit;
  AViewInfo.Transparent := True;
  FillGradientRect(ACanvas, AViewInfo.Bounds, $50F0F0, $009090, False);
  ACanvas.Brush.Style := bsClear;
  ACanvas.Font := csBoldItalic.Font;
  ACanvas.DrawText(AViewInfo.DisplayText, AViewInfo.Bounds, cxAlignCenter);
  ACanvas.Brush.Style := bsSolid;
  ADone := True;
end;

procedure TCustomDrawDemoMainForm.SchedulerCustomDrawGroupSeparator(
  Sender: TObject; ACanvas: TcxCanvas;
  AViewInfo: TcxSchedulerGroupSeparatorCellViewInfo; var ADone: Boolean);
begin
  if not miGroupSeparator.Checked then Exit;
  FillGradientRect(ACanvas, cxRectInflate(AViewInfo.Bounds, 0, 0, 1, 0), $50F0F0, $009090, True);
  ADone := True;
end;

procedure TCustomDrawDemoMainForm.FormDestroy(Sender: TObject);
begin
  inherited;
  ABitmap.Free;
end;

procedure TCustomDrawDemoMainForm.UpdateCustomDraw(Sender: TObject);
begin
  with TMenuItem(Sender) do Checked := not Checked;
  Scheduler.LayoutChanged;
end;

end.

File: Components/DevExpressVCL/Demos/ExpressScheduler/CBuilder/CustomDrawDemo/CustomDrawDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop
#define HDC unsigned int

#include "CustomDrawDemoMain.h"

//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxControls"
#pragma link "cxEdit"
#pragma link "cxGraphics"
#pragma link "cxScheduler"
#pragma link "cxSchedulerCustomControls"
#pragma link "cxSchedulerCustomResourceView"
#pragma link "cxSchedulerDateNavigator"
#pragma link "cxSchedulerDayView"
#pragma link "cxStyles"
#pragma link "DemoBasicMain"
#pragma link "cxSchedulerStorage"
#pragma link "cxFormats"
#pragma link "dxOffice11"

#pragma link "DemoBasicMain"
#pragma link "cxSchedulerGanttView"
#pragma link "cxSchedulerHolidays"
#pragma link "cxSchedulerTimeGridView"
#pragma link "cxSchedulerUtils"
#pragma link "cxSchedulerWeekView"
#pragma link "cxSchedulerYearView"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxLookAndFeels"
#pragma link "cxSchedulerGanttView"
#pragma link "cxSchedulerHolidays"
#pragma link "cxSchedulerTimeGridView"
#pragma link "cxSchedulerUtils"
#pragma link "cxSchedulerWeekView"
#pragma link "cxSchedulerYearView"
#pragma link "cxSchedulerAgendaView"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxLookAndFeels"
#pragma link "cxSchedulerGanttView"
#pragma link "cxSchedulerHolidays"
#pragma link "cxSchedulerRecurrence"
#pragma link "cxSchedulerTimeGridView"
#pragma link "cxSchedulerUtils"
#pragma link "cxSchedulerWeekView"
#pragma link "cxSchedulerYearView"
#pragma resource "*.dfm"

TCustomDrawDemoMainForm *CustomDrawDemoMainForm;

  int W1;
  int H1;
  int C1;
  int C2;
  Graphics::TBitmap *ABitmap;
  unsigned char FDaysOffIndexes[2];

// we do not use dxOffice11.FillGradientRect because of the problem in C++Builder during linking
void __fastcall FillGradientRectEx(TcxCanvas* ACanvas, const TRect &ARect, int AColor1, int AColor2, bool AHorizontal)
{
  int AFirstOffset, ALastOffset, APixelSteps, AColorStepsR, AColorStepsG, AColorStepsB;
  TRect R = ARect;

  if (AHorizontal) {
    AFirstOffset = ARect.Left;
    ALastOffset = ARect.Right - 1;
  }
  else
  {
    AFirstOffset = ARect.Top;
    ALastOffset = ARect.Bottom - 1;
  }
  APixelSteps = ALastOffset - AFirstOffset;
  AColorStepsR = GetRValue(AColor2) - GetRValue(AColor1);
  AColorStepsG = GetGValue(AColor2) - GetGValue(AColor1);
  AColorStepsB = GetBValue(AColor2) - GetBValue(AColor1);

  for (int I = AFirstOffset; I <= ALastOffset; I++) {
    if (AHorizontal) {
      R.Left = I;
      R.Right = I + 1;
    }
	else
    {
      R.Top = I;
      R.Bottom = I + 1;
    }
    ACanvas->Brush->Color = (TColor)RGB(
      GetRValue(AColor1) + MulDiv(I - AFirstOffset, AColorStepsR, APixelSteps),
      GetGValue(AColor1) + MulDiv(I - AFirstOffset, AColorStepsG, APixelSteps),
      GetBValue(AColor1) + MulDiv(I - AFirstOffset, AColorStepsB, APixelSteps));
    ACanvas->FillRect(R, NULL, False);
  }
}

//---------------------------------------------------------------------------
__fastcall TCustomDrawDemoMainForm::TCustomDrawDemoMainForm(TComponent* Owner)
        : TDemoBasicMainForm(Owner)
{
//
}

TcxLookAndFeelKind __fastcall TCustomDrawDemoMainForm::GetDefaultLookAndFeelKind()
{
  return(lfUltraFlat);
}

//---------------------------------------------------------------------------
void __fastcall TCustomDrawDemoMainForm::FormCreate(TObject *Sender)
{
  TDemoBasicMainForm::FormCreate(Sender);
  ABitmap = new Graphics::TBitmap();
  ABitmap->PixelFormat = pf32bit;
  Scheduler->OptionsView->RotateResourceCaptions = false;
  Scheduler->ViewDay->EventShadows = false;
  int I = (int)cxFormatController()->StartOfWeek + (int)(dSunday);
  int J = (int)(dSunday) - I;
  if (J < 0) J = J + 7;
  FDaysOffIndexes[0] = J;
  J = (int)(dSaturday) - I;
  if (J < 0) J = J + 7;
  FDaysOffIndexes[1] = J;
  GenerateRandomEvents(100, true);
}
//---------------------------------------------------------------------------
void __fastcall TCustomDrawDemoMainForm::FormDestroy(TObject *Sender)
{
  delete ABitmap;
}
//---------------------------------------------------------------------------
void __fastcall TCustomDrawDemoMainForm::SchedulerDateNavigatorCustomDrawHeader(
  TObject *Sender, TcxCanvas *ACanvas, TcxSchedulerDateNavigatorMonthHeaderViewInfo *AViewInfo, bool &ADone)
{
  if (!miMonthHeaders->Checked)
    return;
  FillGradientRectEx(ACanvas, AViewInfo->Bounds, 0xFFE0E0, 0xFF8080, false);
  ACanvas->FrameRect(AViewInfo->Bounds, clBlue, 1, TcxBorders()<<bLeft<<bRight<<bTop<<bBottom, False);
  AViewInfo->Transparent = true;
}
//---------------------------------------------------------------------------
void __fastcall TCustomDrawDemoMainForm::SchedulerDateNavigatorCustomDrawDayCaption(
  TObject *Sender, TcxCanvas *ACanvas, TcxSchedulerDateNavigatorDayCaptionViewInfo *AViewInfo, bool &ADone)
{
  if (!miDayCaptions->Checked)
	return;
  if (AViewInfo->Index == FDaysOffIndexes[0] || AViewInfo->Index == FDaysOffIndexes[1])
    ACanvas->Font = csRed->Font;
}
//---------------------------------------------------------------------------
void __fastcall TCustomDrawDemoMainForm::SchedulerDateNavigatorCustomDrawDayNumber(
  TObject *Sender, TcxCanvas *ACanvas, TcxSchedulerDateNavigatorDayNumberViewInfo *AViewInfo, bool &ADone)
{
  if (!miDays->Checked) return;
  int ADay = DayOfWeek(AViewInfo->Date);
  if (AViewInfo->Selected)
    ACanvas->Brush->Color = clAppWorkSpace;
  else
    if (ADay == 1 || ADay == 7)
      ACanvas->Font = csRed->Font;
}
//---------------------------------------------------------------------------
void __fastcall TCustomDrawDemoMainForm::SchedulerCustomDrawDayHeader(
  TObject *Sender, TcxCanvas *ACanvas, TcxSchedulerDayHeaderCellViewInfo *AViewInfo, bool &ADone)
{
  if (!miHeaders->Checked) return;
  AViewInfo->Transparent = true;
  FillGradientRectEx(ACanvas, AViewInfo->Bounds, 0xA0A0A0, 0x707070, false);
  ACanvas->FrameRect(AViewInfo->Bounds, clGray, 1, TcxBorders()<<bLeft<<bRight<<bTop<<bBottom, false);
  ACanvas->Brush->Style = bsClear;
  ACanvas->Font = csItalic->Font;
  if ((dynamic_cast<TcxSchedulerAgendaViewDayHeaderCellViewInfo*>(AViewInfo)) && (Scheduler->ViewAgenda->DayHeaderOrientation == dhoVertical))
  {
	String ADisplayText = AViewInfo->DisplayText;
	int P = Pos(',', ADisplayText);
	ADisplayText = ADisplayText.SubString(1, P - 1) + "\r\n" + ADisplayText.SubString(P + 2, ADisplayText.Length() - P - 1);
	TRect ABounds = cxRectInflate(AViewInfo->Bounds, -2 * cxTextOffset, -2 * cxTextOffset);
	ACanvas->DrawTexT(ADisplayText, ABounds, cxAlignHCenter | cxAlignTop | cxWordBreak);
  }
  else
    ACanvas->DrawTexT(AViewInfo->DisplayText, AViewInfo->Bounds, cxAlignCenter, true);
  ACanvas->Brush->Style = bsSolid;
  ADone = True;
}
//---------------------------------------------------------------------------
void __fastcall TCustomDrawDemoMainForm::SchedulerViewDayCustomDrawRuler(
  TObject *Sender, TcxCanvas *ACanvas, TcxSchedulerTimeRulerCellViewInfo *AViewInfo, bool &ADone)
{
  if (!miTimeRuler->Checked) return;
  AViewInfo->Transparent = true;
  FillGradientRectEx(ACanvas, AViewInfo->Bounds[true], 0x909090, 0xE0E0E0, true);
}
//---------------------------------------------------------------------------
void __fastcall TCustomDrawDemoMainForm::SchedulerViewDayCustomDrawContainer(
  TObject *Sender, TcxCanvas *ACanvas, TcxSchedulerContainerCellViewInfo *AViewInfo, bool &ADone)
{
  if (!miContainer->Checked || AViewInfo->Selected) return;
  AViewInfo->Transparent = true;
  FillGradientRectEx(ACanvas, AViewInfo->Bounds, 0xC0E0C0, 0x70A070, false);
}
//---------------------------------------------------------------------------
void __fastcall TCustomDrawDemoMainForm::SchedulerCustomDrawEvent(
  TObject *Sender, TcxCanvas *ACanvas, TcxSchedulerEventCellViewInfo *AViewInfo, bool &ADone)
{
  if (!miEvents->Checked) return;
  AViewInfo->Transparent = true;
  if (dynamic_cast<TcxSchedulerAgendaViewEventCellViewInfo*>(AViewInfo))
    FillGradientRectEx(ACanvas, AViewInfo->Bounds, clWhite, dxGetMiddleRGB(clBlack, AViewInfo->Color, 10), false);
  else
    FillGradientRectEx(ACanvas, AViewInfo->Bounds, 0xA0A0A0, AViewInfo->Color, true);
}
//---------------------------------------------------------------------------
void __fastcall TCustomDrawDemoMainForm::UpdateCustomDraw(TObject *Sender)
{
  ((TMenuItem*)Sender)->Checked = !((TMenuItem*)Sender)->Checked;
  Scheduler->LayoutChanged();
}
//---------------------------------------------------------------------------
void __fastcall TCustomDrawDemoMainForm::SchedulerCustomDrawContent(
  TObject *Sender, TcxCanvas *ACanvas, TcxSchedulerContentCellViewInfo *AViewInfo, bool &ADone)
{
  if (!miContent->Checked) return;
  AViewInfo->Transparent = true;
  if (dynamic_cast<TcxSchedulerAgendaViewFreeDayCellViewInfo*>(AViewInfo))
    FillGradientRectEx(ACanvas, AViewInfo->Bounds, clWhite, dxGetMiddleRGB(clBlack, AViewInfo->Color, 10), false);
  else
    FillGradientRectEx(ACanvas, AViewInfo->Bounds, AViewInfo->Color, dxGetMiddleRGB(AViewInfo->Color, clBlack, 75), true);
}
//---------------------------------------------------------------------------
void __fastcall TCustomDrawDemoMainForm::SchedulerCustomDrawResourceHeader(
  TObject *Sender, TcxCanvas *ACanvas, TcxSchedulerHeaderCellViewInfo *AViewInfo, bool &ADone)
{
  if (!miResources->Checked) return;
  AViewInfo->Transparent = True;
  FillGradientRectEx(ACanvas, AViewInfo->Bounds, 0x50F0F0, 0x009090, false);
  ACanvas->Brush->Style = bsClear;
  ACanvas->Font = csBoldItalic->Font;
  ACanvas->DrawTexT(AViewInfo->DisplayText, AViewInfo->Bounds, cxAlignCenter, true);
  ACanvas->Brush->Style = bsSolid;
  ADone = True;
}
//---------------------------------------------------------------------------
void __fastcall TCustomDrawDemoMainForm::SchedulerCustomDrawGroupSeparator(
  TObject *Sender, TcxCanvas *ACanvas, TcxSchedulerGroupSeparatorCellViewInfo *AViewInfo, bool &ADone)
{
  if (!miGroupSeparator->Checked) return;
  FillGradientRectEx(ACanvas, AViewInfo->Bounds, 0x50F0F0, 0x009090, true);
  ADone = True;
}
//---------------------------------------------------------------------------
void __fastcall TCustomDrawDemoMainForm::SchedulerDateNavigatorCustomDrawContent(
  TObject *Sender, TcxCanvas *ACanvas, TcxSchedulerDateNavigatorMonthContentViewInfo *AViewInfo, bool &ADone)
{
  TColor AColor;
  if (!miDNContent->Checked) return;
  if (AViewInfo->Month >= 3 && AViewInfo->Month <= 11)
  {
    if (AViewInfo->Month <= 5)
      AColor = (TColor)0xD0FFD0;
    else
      if (AViewInfo->Month <= 8)
        AColor = (TColor)0xD0D0FF;
      else
        if (AViewInfo->Month <= 11)
          AColor = (TColor)0xD0FFFF;
  }
  else
    AColor = (TColor)0xFFE7E7;
  TRect R = AViewInfo->Bounds;
  ACanvas->Brush->Color = AColor;
  ACanvas->FillRect(R, NULL, false);
  ACanvas->Font->Height = R.Height();
  ACanvas->Font->Color = (TColor)dxGetMiddleRGB(AColor, clBlack, 85);
  ACanvas->DrawTexT(IntToStr(AViewInfo->Month), R, cxAlignCenter, true);
  ACanvas->Font = AViewInfo->ViewParams.Font;
  AViewInfo->Transparent = true;
  ADone = True;
}
//---------------------------------------------------------------------------

